bundles-lib/src/main/java/org/apache/metron/bundles/BundleClassLoaders.java

```diff
@@ -0,0 +1,353 @@
+
+    private InitContext(
+        final List<FileObject> extensionDirs,
+        final Map<String, Bundle> bundles) {
+      this.extensionDirs = extensionDirs;
+      this.bundles = bundles;
+    }
+  }
+
+  private BundleClassLoaders() {
+  }
+
+  /**
+   * @return The singleton instance of the BundleClassLoaders
+   */
+  public static BundleClassLoaders getInstance() {
+    BundleClassLoaders result = bundleClassLoaders;
+    if (result == null) {
+      synchronized (BundleClassLoaders.class) {
+        result = bundleClassLoaders;
+        if (result == null) {
+          bundleClassLoaders = result = new BundleClassLoaders();
+        }
+      }
+    }
+    return result;
+  }
+
+  public static void reset() {
+    getInstance().unInit();
+  }
+
+  private void unInit() {
+    synchronized (this) {
+      initContext = null;
+    }
+  }
+
+  /**
+   * Initializes and loads the BundleClassLoaders. This method must be called before the rest of the
+   * methods to access the classloaders are called and it can be safely called any number of times
+   * provided the same framework and extension working dirs are used.
+   *
+   * @param fileSystemManager the FileSystemManager
+   * @param extensionsDirs where to find extension artifacts
+   * @param props BundleProperties
+   * @throws FileSystemException if any issue occurs while exploding bundle working directories.
+   * @throws java.lang.ClassNotFoundException if unable to load class definition
+   * @throws IllegalStateException already initialized with a given pair of directories cannot
+   * reinitialize or use a different pair of directories.
+   */
+  public void init(final FileSystemManager fileSystemManager, final List<FileObject> extensionsDirs,
+      BundleProperties props)
+      throws FileSystemException, ClassNotFoundException, URISyntaxException {
+    if (extensionsDirs == null || fileSystemManager == null) {
+      throw new NullPointerException("cannot have empty arguments");
+    }
+    InitContext ic = initContext;
+    if (ic == null) {
+      synchronized (this) {
+        ic = initContext;
+        if (ic == null) {
+          initContext = ic = load(fileSystemManager, extensionsDirs, props);
+        }
+      }
+    } else {
+      boolean matching = CollectionUtils
+          .isEqualCollection(initContext.extensionDirs, extensionsDirs);
+      if (!matching) {
+        throw new IllegalStateException(
+            "Cannot reinitialize and extension/framework directories cannot change");
+      }
+    }
+  }
+
+  /**
+   * Should be called at most once.
+   */
+  private InitContext load(final FileSystemManager fileSystemManager,
+      final List<FileObject> extensionsDirs, BundleProperties props)
+      throws FileSystemException, ClassNotFoundException, URISyntaxException {
+    // get the system classloader
+    final ClassLoader systemClassLoader = ClassLoader.getSystemClassLoader();
+
+    // find all bundle files and create class loaders for them.
```

Comment:
The only way this would be a problem, is if the 'loader'/shaded jar is not the parent for the correct extension type.  I don't think that can happen.  I just needed to think it through.

So, even in the storm single JVM, when loading topology jars it would be:

[system classloader] -> topology A jar Classloader -> Bundle ClassLoader
[system classloader] -> topology B jar Classloader -> Bundle ClassLoader

Each should be fine with their own environments etc.